File: spryntax/website_static_build_tools/raw_alg_builder/algs-py/dlink.py
# Python Program for traversal of a doubly linked list
class Node:
	def __init__(self, data):
		# Initialize a new node with data, previous, and next pointers
		self.data = data
		self.next = None
		self.prev = None


def traverse(head):
	# Traverse the doubly linked list and print its elements
	current = head
	while current:
	  # Print current node's data
		print(current.data, end=" <-> ")
		# Move to the next node
		current = current.next
	print("None")


def insert_at_beginning(head, data):
	# Insert a new node at the beginning of the doubly linked list
	new_node = Node(data)
	new_node.next = head
	if head:
		head.prev = new_node
	return new_node


# Driver Code
head = None
head = insert_at_beginning(head, 4)
head = insert_at_beginning(head, 3)
head = insert_at_beginning(head, 2)
head = insert_at_beginning(head, 1)

# To traverse and print the nodes:
traverse(head)

File: spryntax/website_static_build_tools/alg_creator/algs-c/dlink.c
#include <stdio.h>
#include <stdlib.h>

// Define the Node structure
struct Node {
	// Data stored in the node
	int data;
	// Pointer to the next node
	struct Node* next;
	// Pointer to the previous node
	struct Node* prev;
};

// Function to create a new node
struct Node* createNode(int data) {
	struct Node* newNode = 
	  (struct Node*)malloc(sizeof(struct Node));
	newNode->data = data;
	newNode->next = NULL;
	newNode->prev = NULL;
	return newNode;
}

// Function to traverse the doubly linked list 
// in forward direction
void forwardTraversal(struct Node* head) {
  
	// Start traversal from the head of the list
	struct Node* curr = head;

	// Continue until the current node is not
	// null (end of list)
	while (curr != NULL) {
	  
		// Output data of the current node
		printf("%d ", curr->data);
	  
		// Move to the next node
		curr = curr->next;
	}

	// Print newline after traversal
	printf("\n");
}

// Function to traverse the doubly linked list 
// in backward direction
void backwardTraversal(struct Node* tail) {
  
	// Start traversal from the tail of the list
	struct Node* curr = tail;

	// Continue until the current node is not 
	// null (end of list)
	while (curr != NULL) {
	  
		// Output data of the current node
		printf("%d ", curr->data);
	  
		// Move to the previous node
		curr = curr->prev;
	}

	// Print newline after traversal
	printf("\n");
}

int main() {
  
	// Sample usage of the doubly linked list and 
	// traversal functions
	struct Node* head = createNode(1);
	struct Node* second = createNode(2);
	struct Node* third = createNode(3);

	head->next = second;
	second->prev = head;
	second->next = third;
	third->prev = second;

	printf("Forward Traversal:\n");
	forwardTraversal(head);

	printf("Backward Traversal:\n");
	backwardTraversal(third);

	// Free memory allocated for nodes
	free(head);
	free(second);
	free(third);

	return 0;
}